### PR TITLE
add visreader

### PR DESCRIPTION
**Pull request by Yancey0623**

### CODE REVIEW

File: fluid/PaddleCV/image_classification/dist_train/README.md [hunk 1]
```diff
@@ -86,6 +86,7 @@ The default resnet50 distributed training config is based on this paper: https:/
   we run several batches forward and backward computation, then merge the gradients and send to pserver for
   optimization, we use different batch norm mean and variance variable in each repeat so that adding repeats
   behaves the same as adding more GPUs.
+- use [VisionTools DataReader](https://github.com/PaddlePaddle/VisionTools/tree/master/datareader) `--visreader` can achieve better performance to process images.
```
Comment (typhoonzero) [inline fluid/PaddleCV/image_classification/dist_train/README.md]: Add `requirements.txt` to install: paddlepaddle-gpu and visiontool?

File: fluid/PaddleCV/image_classification/dist_train/dist_train.py [hunk 2]
```diff
@@ -66,6 +66,7 @@ def parse_args():
     add_arg('async_mode',         bool, False,              "Async distributed training, only for pserver mode.")
     add_arg('reduce_strategy',    str,  "allreduce",        "Choose from reduce or allreduce.")
     add_arg('skip_unbalanced_data', bool, False,            "Skip data not if data not balanced on nodes.")
+    add_arg('visreader',            bool, False,            "Whether to use high performance VisTool reader.")
```
Comment (typhoonzero) [inline fluid/PaddleCV/image_classification/dist_train/dist_train.py]: use_visiontool

File: fluid/PaddleCV/image_classification/dist_train/dist_train.py [hunk 3]
```diff
@@ -94,6 +95,40 @@ def prepare_reader(is_train, pyreader, args, pass_id=0):
...
+        if len(obj['label']) >= 4:
+            label = int(obj['label'][3])  # class id
+        else:
+            label = int(obj['label'][2])
+        return obj['image'], label
+    pre_maps = [_parse_kv]
+    worker_args = {}
+    worker_args['cpp_xmap'] = True
+    worker_args['use_process'] = False
+    worker_args['worker_num'] = 16
```
Comment (typhoonzero) [inline fluid/PaddleCV/image_classification/dist_train/dist_train.py]: This should be configurable

File: fluid/PaddleCV/image_classification/dist_train/dist_train.py [hunk 4]
```diff
@@ -103,17 +138,23 @@ def build_program(is_train, main_prog, startup_prog, args):
     device_num_per_worker = get_device_num()
     with fluid.program_guard(main_prog, startup_prog):
         pyreader = fluid.layers.py_reader(
-            capacity=16,
+            capacity=args.batch_size,
             shapes=([-1] + image_shape, (-1, 1)),
-            dtypes=('float32', 'int64'),
+            dtypes=('uint8', 'int64'),
```
Comment (typhoonzero) [inline fluid/PaddleCV/image_classification/dist_train/dist_train.py:148]: Do we need add options to enable uint8 reader?

File: fluid/PaddleCV/image_classification/dist_train/dist_train.py [hunk 5]
```diff
@@ -103,17 +138,23 @@ def build_program(is_train, main_prog, startup_prog, args):
...
             name="train_reader" if is_train else "test_reader",
             use_double_buffer=True)
         with fluid.unique_name.guard():
             image, label = fluid.layers.read_file(pyreader)
+            # normalize image on GPU can achieve better performance
+            cast = fluid.layers.cast(image, "float32")
```
Comment (typhoonzero) [inline fluid/PaddleCV/image_classification/dist_train/dist_train.py]: should cast to fp16 or fp32 here.

File: fluid/PaddleCV/image_classification/dist_train/dist_train.py [hunk 6]
```diff
@@ -103,17 +138,23 @@ def build_program(is_train, main_prog, startup_prog, args):
...
+            img_mean = fluid.layers.create_global_var([3, 1, 1], 0.0, "float32", name="img_mean", persistable=True)
+            img_std = fluid.layers.create_global_var([3, 1, 1], 0.0, "float32", name="img_std", persistable=True)
+            t1 = fluid.layers.elementwise_sub(cast / 255.0, img_mean, axis=1)
+            t2 = fluid.layers.elementwise_div(t1, img_std, axis=1)
```
Comment (typhoonzero) [inline fluid/PaddleCV/image_classification/dist_train/dist_train.py]: t1, t2 are bad names.

File: fluid/PaddleCV/image_classification/dist_train/dist_train.py [hunk 7]
```diff
@@ -130,7 +171,7 @@ def build_program(is_train, main_prog, startup_prog, args):
                 if os.getenv("FLAGS_selected_gpus"):
                     # in multi process mode, "trainer_count" will be total devices
                     # in the whole cluster, and we need to scale num_of nodes.
-                    end_lr *= device_num_per_worker
+                    end_lr /= device_num_per_worker
```
Comment (typhoonzero) [inline fluid/PaddleCV/image_classification/dist_train/dist_train.py:180]: Thanks for this fix

File: fluid/PaddleCV/image_classification/dist_train/dist_train.py [hunk 8]
```diff
@@ -261,11 +312,14 @@ def train_parallel(args):
     for pass_id in range(args.num_epochs):
         num_samples = 0
         start_time = time.time()
-        batch_id = 1
-        # use pass_id+1 as per pass global shuffle for distributed training
-        prepare_reader(True, train_pyreader, args, pass_id + 1)
-        train_pyreader.start()
-        while True:
+        if args.visreader and pass_id == 0:
+            # vistool reader is ulimited data, don't need restart pyreader
```
Comment (typhoonzero) [inline fluid/PaddleCV/image_classification/dist_train/dist_train.py]: be ware that vision tool do not support global shuffle right now

File: fluid/PaddleCV/image_classification/dist_train/dist_train.py [hunk 9]
```diff
@@ -221,8 +271,21 @@ def train_parallel(args):
...
+        var = fluid.global_scope().find_var(vname)
+        if args.fp16:
+            var.get_tensor().set(np_tensor.view(np.uint16), place)
+        else:
+            var.get_tensor().set(np_tensor, place)
+
+
     strategy = fluid.ExecutionStrategy()
     strategy.num_threads = args.num_threads
+    strategy.num_iteration_per_drop_scope = 30
```
Comment (guru4elephant) [inline fluid/PaddleCV/image_classification/dist_train/dist_train.py]: this option seems not friendly to users. Do we have any other configuration method for this?

File: fluid/PaddleCV/image_classification/dist_train/dist_train.py [hunk 10]
```diff
@@ -65,11 +65,17 @@ def parse_args():
     add_arg('split_var',          bool, True,               "Split params on pserver.")
     add_arg('async_mode',         bool, False,              "Async distributed training, only for pserver mode.")
     add_arg('reduce_strategy',    str,  "allreduce",        "Choose from reduce or allreduce.")
-    add_arg('skip_unbalanced_data', bool, False,            "Skip data not if data not balanced on nodes.")
+    add_arg('skip_unbalanced_data', bool,   False,          "Skip data not if data not balanced on nodes.")
```
Comment (guru4elephant) [inline fluid/PaddleCV/image_classification/dist_train/dist_train.py]: default = True is more friendly?

File: fluid/PaddleCV/image_classification/reader.py [hunk 11]
```diff
@@ -96,7 +96,7 @@ def random_color(img, lower=0.5, upper=1.5):
     return img
 
 
-def process_image(sample, mode, color_jitter, rotate):
+def process_image(sample, mode, color_jitter, rotate, normalize):
```
Comment (guru4elephant) [inline fluid/PaddleCV/image_classification/reader.py]: why not set normalize=True as default option?

**Review by guru4elephant — COMMENTED** [review]
the efficiency of mp_mode should be explained through an experimental result table in README.md. It's ok to show results ran on internal clusters I think, since the key point is to make users know that mp_mode is effective.

File: fluid/PaddleCV/image_classification/dist_train/README.md [hunk 12]
```diff
@@ -86,6 +86,9 @@ The default resnet50 distributed training config is based on this paper: https:/
   we run several batches forward and backward computation, then merge the gradients and send to pserver for
   optimization, we use different batch norm mean and variance variable in each repeat so that adding repeats
   behaves the same as adding more GPUs.
+- use [VisionTools DataReader](https://github.com/PaddlePaddle/VisionTools/tree/master/datareader) `--use_visiontool True` 
+  can achieve better performance to pre-process images. You need to install visiontool on your host before enable it by:
+  `pip install -r requiremenets.txt`.
```
Comment (qingqing01) [inline fluid/PaddleCV/image_classification/dist_train/README.md:91]: 这个能同时作用到单机上吗？

Comment (Yancey0623) [reply fluid/PaddleCV/image_classification/dist_train/README.md:91]: 可以的。

File: fluid/PaddleCV/image_classification/reader.py (hunk 11 above)
Comment (Yancey0623) [reply fluid/PaddleCV/image_classification/reader.py]: norm image in graph can get better performance.

File: fluid/PaddleCV/image_classification/dist_train/dist_train.py (hunk 10 above)
Comment (Yancey0623) [reply fluid/PaddleCV/image_classification/dist_train/dist_train.py]: Will delete this arg, `skip_unbalanced_data` is a deprecated API in pyreader.

File: fluid/PaddleCV/image_classification/dist_train/dist_train.py (hunk 9 above)
Comment (Yancey0623) [reply fluid/PaddleCV/image_classification/dist_train/dist_train.py]: Done.

**Review by typhoonzero — APPROVED** [review]
